While loop

junrushao opened the pull request.

zheng-da committed:
Test input a graph.

zheng-da committed:
Update foreach to execute the subgraph.

zheng-da committed:
print inputs/outputs in foreach.

zheng-da committed:
Remove print.

zheng-da committed:
add test code for foreach.

zheng-da committed:
exec foreach outside the engine.

zheng-da committed:
Implements forward of foreach.

zheng-da committed:
Add support for variable numbers of inputs and outputs.

zheng-da committed:
Add a python wrapper for foreach.

zheng-da committed:
Fix the order of inputs.

zheng-da committed:
add test with lstm.

zheng-da committed:
hide C version of foreach.

zheng-da committed:
fix a bug temporarily.

zheng-da committed:
Test free variables.

zheng-da committed:
change for the new interface of InputGraph attribute.

zheng-da committed:
Add attribute to the subgraph.

zheng-da committed:
Handle free variables.

zheng-da committed:
Get all input symbols of a subgraph.

zheng-da committed:
Fix shape, dtype and storage inference.

zheng-da committed:
reorganize the output of foreach.

zheng-da committed:
Add a gluon RNN unroll with symbol foreach.

zheng-da committed:
print unnecessary print.

zheng-da committed:
have imperative and symbolic foreach.

zheng-da committed:
Fix an error after moving foreach.

zheng-da committed:
Fix imperative foreach

zheng-da committed:
Fix a minor problem.

zheng-da committed:
Use CachedOp to execute subgraph.

zheng-da committed:
update TODO.

zheng-da committed:
make foreach op use FStatefulComputeEx.

TODO we need to change stateful executor to handle subgraph.

zheng-da committed:
Add backward.

zheng-da committed:
Fix bugs.

zheng-da committed:
enable backward test in lstm.

zheng-da committed:
Fix a bug in foreach backward for free variables.

zheng-da committed:
change for the new CachedOp.

zheng-da committed:
Detect the backward computation.

zheng-da committed:
Fix bugs in foreach.

zheng-da committed:
fix tests.

zheng-da committed:
update tests.

zheng-da committed:
check state shape.

zheng-da committed:
enable nested foreach.

zheng-da committed:
remove print.

zheng-da committed:
fix a bug in test.

zheng-da committed:
handle infer storage type for backward.

zheng-da committed:
address comments.

zheng-da committed:
address comments.

zheng-da committed:
move some common functions out.

zheng-da committed:
address comments.

zheng-da committed:
fix lint.

zheng-da committed:
Fix lint.

zheng-da committed:
add doc.

zheng-da committed:
undo modification in imperative.h

zheng-da committed:
add doc and remove example code.

zheng-da committed:
fix lint.

zheng-da committed:
fix lint.

zheng-da committed:
Fix lint.

zheng-da committed:
make nd.foreach and sym.foreach consistent.

zheng-da committed:
fix compile error.

zheng-da committed:
address comments.

zheng-da committed:
update.

zheng-da committed:
check for loop only works for dense arrays.

zheng-da committed:
move control flow op out of nn/

zheng-da committed:
fix include.

zheng-da committed:
add a test in gluon.

zheng-da committed:
work for GPU.

zheng-da committed:
small fix.

zheng-da committed:
remove subgraph_name

zheng-da committed:
create loop state for reuse in the future.

zheng-da committed:
move code.

zheng-da committed:
Revert "remove subgraph_name"

This reverts commit 977f5624ad0b0dedb9dcb8629f975afc56bb1e1a.

zheng-da committed:
cut graph.

zheng-da committed:
rename new var nodes.

zheng-da committed:
Fix tests.

Commit:
Fix bugs caused by ctypes (#29)

Commit:
Add save/load json in testcases for foreach (#30)

zheng-da committed:
support subgraph in stateful executor.

zheng-da committed:
Fix compilation.

zheng-da committed:
fix a bug when a subgraph has variable nodes.

zheng-da committed:
Fix a bug of getting symbols.

zheng-da committed:
copy var nodes.

zheng-da committed:
Fix getting op states.

zheng-da committed:
fix lint error.

zheng-da committed:
address comments.

zheng-da committed:
fix lint error.

zheng-da committed:
simplify the execution of subgraph in the main thread.

zheng-da committed:
fix lint error.

zheng-da committed:
avoid waiting for computation in each iteration.

zheng-da committed:
reuse cached op for inference.

zheng-da committed:
share memory across mini-batches.

zheng-da committed:
reuse memory.

reuse memory between iterations in inference.
reuse memory between mini-batches in training.

zheng-da committed:
add tests for multiple batches.

zheng-da committed:
remove entry.

zheng-da committed:
add benchmark for foreach.

zheng-da committed:
benchmark large batch size.

zheng-da committed:
Fix the benchmark for GPU.

zheng-da committed:
address comments.

zheng-da committed:
update shape/dtype/storage inference.

zheng-da committed:
update contrib API docs.

zheng-da committed:
support nested foreach.

zheng-da committed:
use a single CachedOp for all iterations.

zheng-da committed:
use large dim.

zheng-da committed:
update benchmark.

zheng-da committed:
update benchmark.

zheng-da committed:
update benchmark.

zheng-da committed:
update benchmark.

zheng-da committed:
return symbol arrays correctly in MXSymbolCutSubgraph.

zheng-da committed:
return symbol arrays in MXSymbolGetInputSymbols.

zheng-da committed:
fix lint error.

zheng-da committed:
use cachedop to infer storage in backward.

junrushao committed:
Add while_loop

junrushao committed:
Add while_loop

zheng-da committed:
fix scala API.

zheng-da committed:
update comments.

junrushao committed:
...

junrushao committed:
...

zheng-da committed:
fix scala.

zheng-da committed:
fix test.

zheng-da committed:
fix attribute name.

zheng-da committed:
move benchmark.

zheng-da commented on python/mxnet/ndarray/contrib.py:
```diff
@@ -192,3 +192,126 @@ def check_input(inputs, in_type, msg):
+    2) optionally outputs for each step
+    The signature of ``fun'' can be:
+       * def func(loop_vars): new_loop_vars
+    or * def func(loop_vars): (output, new_loop_vars)
```
new_loop_vars should have the same number of elements as loop_vars. The shape and data type of NDArrays in new_loop_vars should match the ones in loop_vars.

junrushao replied:
solved

zheng-da commented on python/mxnet/ndarray/contrib.py:
```diff
@@ -192,3 +192,126 @@ def check_input(inputs, in_type, msg):
+
+    ``loop_vars'' is a list of NDArrays that represent loop variables.
+
+    ``max_iterations'' is a python or NDArray scalar that defines the maximal number of iterations.
```
why max_iterations should be a NDArray scalar?

junrushao replied:
It could be a python scalar, or a NDArray scalar, or a numpy ndarray scalar, _to_python_scalar below will convert it to a python scalar

junrushao replied:
solved

zheng-da commented on python/mxnet/ndarray/contrib.py:
```diff
@@ -192,3 +192,126 @@ def check_input(inputs, in_type, msg):
+        to the given type
+        """
+        if isinstance(inputs, ndarray.NDArray):
+            inputs = inputs.asscalar()
```
do you need to cast the type here as well?

junrushao replied:
Nope. inputs will be cast to the given `type` inside the try block so that any cast error will be caught and reported

junrushao replied:
solved

zheng-da commented on python/mxnet/ndarray/contrib.py:
```diff
@@ -192,3 +192,126 @@ def check_input(inputs, in_type, msg):
+    TODO(Junru): review the documentation, it has been out-of-date.
+    Corner case #1: what if cond is always false? Do we return "outputs"?
+    """
+    def _to_python_type(inputs, type, name):
```
I suppose this function is to cast to Python scalar? If so, please rename it to make it more explicit.

junrushao replied:
solved

zheng-da commented on python/mxnet/ndarray/contrib.py:
```diff
@@ -192,3 +192,126 @@ def check_input(inputs, in_type, msg):
+            result = tuple(result)
+        if not isinstance(result, tuple):
+            raise ValueError("Invalid return type of func: %s" % (type(result).__name__))
+        if len(result) == 2 and (isinstance(result[1], list) or isinstance(result[1], tuple) or len(loop_vars) == 1):
```
why the last one isn't len(result[1])?

junrushao replied:
solved

zheng-da commented on python/mxnet/ndarray/contrib.py:
```diff
@@ -192,3 +192,126 @@ def check_input(inputs, in_type, msg):
+         and "func: loop_vars -> (step_output, new_loop_vars)"
+        into "func: loop_vars -> (None or tuple of step_outputs, tuple of new_loop_vars)
+        """
+        result = func(*loop_vars)
```
should func accept the iteration number? otherwise, how to get a slice from an NDArray when running RNN?

junrushao replied:
Now users can use the `take` operator for indexing along the axis=0. After #11326 is merge, users are able to index along an arbitrary axis

junrushao replied:
marking this as solved

zheng-da commented on python/mxnet/symbol/contrib.py:
```diff
@@ -296,3 +296,142 @@ def check_data(inputs, in_type, msg):
+        result = cond(*loop_vars)
+        if not isinstance(result, Symbol):
+            raise ValueError("Return of cond must be a Symbol")
+        return [], (result, )
```
why does it return a tuple like this? should it be a symbol that represents a scalar?

junrushao replied:
to be consistent with _func_wrapper

zheng-da commented on python/mxnet/symbol/contrib.py:
```diff
@@ -296,3 +296,142 @@ def check_data(inputs, in_type, msg):
+        return step_output, new_loop_vars
+
+    def _create_subgraph(graph_vars, graph_func, subgraph_name):
+        with AttrScope(subgraph_name=subgraph_name):
```
i just modified the backend implementation. now we need to use `__subgraph_name__`.

junrushao replied:
cool

junrushao replied:
I change this to `AttrScope(__subgraph_name__=subgraph_name)`. Is this as expected?

zheng-da commented on python/mxnet/symbol/contrib.py:
```diff
@@ -296,3 +296,142 @@ def check_data(inputs, in_type, msg):
+            graph = symbol.Group(outputs)
+        return graph, num_out_data, num_outputs
+
+    def _union_inputs(*graphs):
```
could you add comments for this function to explain what it does?

junrushao replied:
solved

zheng-da commented on src/operator/control_flow.cc:
```diff
@@ -363,6 +363,393 @@ ForeachGradient(const nnvm::NodePtr& n, const std::vector<nnvm::NodeEntry>& ogra
+  for (size_t i = 0; i < (size_t) params.num_out_data; i++)
+    CHECK_EQ(params.max_iterations, outputs[i].shape()[0]);
+  for (const auto &arr : outputs)
+    CHECK_EQ(arr.storage_type(), kDefaultStorage) << "The for operator doesn't support the sparse format";
```
`for` => `while`

junrushao replied:
solved

zheng-da commented on src/operator/control_flow.cc:
```diff
@@ -363,6 +363,393 @@ ForeachGradient(const nnvm::NodePtr& n, const std::vector<nnvm::NodeEntry>& ogra
+  int num_out_data;
+  int max_iterations;
+  nnvm::Tuple<dim_t> cond_input_locs;
+  nnvm::Tuple<dim_t> cond_var_locs;
```
please comment the difference between these two fields.

junrushao replied:
solved

zheng-da commented on src/operator/control_flow.cc:
```diff
@@ -363,6 +363,393 @@ ForeachGradient(const nnvm::NodePtr& n, const std::vector<nnvm::NodeEntry>& ogra
+  nnvm::Tuple<dim_t> cond_input_locs;
+  nnvm::Tuple<dim_t> cond_var_locs;
+  nnvm::Tuple<dim_t> func_input_locs;
+  nnvm::Tuple<dim_t> func_var_locs;
```
same here.

junrushao replied:
solved

zheng-da commented on src/operator/control_flow.cc:
```diff
@@ -363,6 +363,393 @@ ForeachGradient(const nnvm::NodePtr& n, const std::vector<nnvm::NodeEntry>& ogra
+    // the SyncCopyToCPU here might be problematic
+    state.cond_op->Forward(nullptr, cond_input_ptr, cond_output_ptr);
+    NDArray *scalar = cond_output_ptr[0];
+    CPUStorage storage(scalar->shape().Size());
```
the size may not be correct. it should be `scalar->shape().Size() * elem_size`

junrushao replied:
solved

zheng-da commented on src/operator/control_flow.cc:
```diff
@@ -363,6 +363,393 @@ ForeachGradient(const nnvm::NodePtr& n, const std::vector<nnvm::NodeEntry>& ogra
+    NDArray *scalar = cond_output_ptr[0];
+    CPUStorage storage(scalar->shape().Size());
+    scalar->SyncCopyToCPU(storage.data, storage.size);
+    return (bool) storage.data[0];
```
we need to use `mshadow_type_switch`

junrushao replied:
right, i am about to do that

junrushao replied:
solved

zheng-da commented on src/operator/control_flow.cc:
```diff
@@ -363,6 +363,393 @@ ForeachGradient(const nnvm::NodePtr& n, const std::vector<nnvm::NodeEntry>& ogra
+    // func_outputs[num_out_data: ] are new_loop_vars, need to allocate new memory
+    for (size_t i = params.num_out_data; i < outputs.size(); ++i) {
+      size_t j = i - params.num_out_data;
+      func_outputs[i] = NDArray(inputs[j].shape(), inputs[j].ctx(), true, inputs[j].dtype());
```
new NDArrays are created in every iteration. we need to avoid creating new NDArrays in the inference.

junrushao replied:
mark it as TODO par our discussion

zheng-da commented on src/operator/control_flow.cc:
```diff
@@ -363,6 +363,393 @@ ForeachGradient(const nnvm::NodePtr& n, const std::vector<nnvm::NodeEntry>& ogra
+                                  const OpContext& ctx,
+                                  const std::vector<NDArray>& inputs,
+                                  const std::vector<OpReqType>& req,
+                                  const std::vector<NDArray>& outputs) {
```
Since while loop always returns NDArrays whose size is max_iteration, we need a way to communicate with users where the valid data is in the output array.

junrushao replied:
state.n_iterations provides the actual number of iterations we have done. To communicate with users, we could expose it to a NDArray scalar, or something else.

Also, users could use an NDArray scalar like `i` to indicate this.

junrushao replied:
Marked as skipped par our discussion

zheng-da commented on src/operator/control_flow.cc:
```diff
@@ -363,6 +363,393 @@ ForeachGradient(const nnvm::NodePtr& n, const std::vector<nnvm::NodeEntry>& ogra
+          // a var
+          igrads[i] = (step == 0)
+                    ? outputs[i]
+                    : NDArray(outputs[i].shape(), outputs[i].ctx(), true, outputs[i].dtype());
```
same here. we can potentially avoid memory allocation here.

junrushao replied:
Marked as TODO as per our discussion

zheng-da commented on src/operator/control_flow.cc:
```diff
@@ -363,6 +363,393 @@ ForeachGradient(const nnvm::NodePtr& n, const std::vector<nnvm::NodeEntry>& ogra
+    for (size_t i = params.num_out_data; i < outputs.size(); ++i) {
+      size_t j = i - params.num_out_data;
+      CHECK_EQ(func_inputs[j].shape(), func_outputs[i].shape());
+      func_inputs[j] = func_outputs[i];
```
func_inputs have two types of inputs. Some of them are loop vars and the others are referenced in the Python closure. Do you keep the variables in the closure in the beginning of the vector? The output loop vars have the same order as the input loop vars?

junrushao replied:
Yes, please refer to my comments in _union_inputs

zheng-da commented on src/operator/control_flow.cc:
```diff
@@ -363,6 +363,393 @@ ForeachGradient(const nnvm::NodePtr& n, const std::vector<nnvm::NodeEntry>& ogra
+    // copy done, call InferShape
+    g.attrs["shape"] = std::make_shared<dmlc::any>(std::move(shapes));
+    g = exec::InferShape(std::move(g));
+    if (g.GetAttr<size_t>("shape_num_unknown_nodes") != 0) {
```
I think you need to copy the inferred shapes back first before returning. Shape inference takes multiple rounds. each round propagates shape info. having unknown shapes doesn't mean the shape inference fails.

junrushao replied:
Please allow me to summarize it:

Before shape inference, should
a) check consistency and possibly assign: `loop_vars` and `new_loop_vars`

After shape inference in a subgraph, should
b) check consistency and possibly assign: `loop_vars` and `new_loop_vars`
c) check consistency and possibly assign: `subg_in` (shape of inputs of the subgraph) and `in_shape` (shape of inputs of the while_loop)

I guess a) could be omitted when a correct control flow is provided, because type inference is done in many rounds, and c) will cover a) finally. But it is still something necessary because we need to report it if users made any mistake.

This is not limited to shape inference, I should also do it in type inference etc.

TODOs:
- [x] shape inference
- [x] type inference
- [x] storage type inference

junrushao1994 replied:
TODO List:
- [ ] shape inference
- [ ] type inference
- [ ] storage type inference

junrushao replied:
solved

junrushao committed:
fix

junrushao committed:
fix

junrushao committed:
Fix input ordering

junrushao committed:
...

junrushao committed:
...

junrushao committed:
Fix bug when some input is output

junrushao committed:
Fix testcases

junrushao committed:
Pass all existing testcases

junrushao committed:
Remove useless cond_var_locs

junrushao committed:
Address warning

zheng-da committed:
fix the mapping of operator inputs/outputs and subgraph inputs/outputs.

zheng-da committed:
add tests for dtype/shape inference.

zheng-da committed:
reorganize tests.

zheng-da committed:
fix a bug of cutting NodeEntry.

When two node entries refer to the same output of a node, we should
create only one var node for these two node entries.

zheng-da committed:
fix lint error.

junrushao committed:
Draft API doc for mx.nd.contrib.while_loop

junrushao committed:
Fix issues in mx.nd.contrib.while_loop

junrushao committed:
Complete API mx.[sym|nd].contrib.while_loop

junrushao committed:
Add a few TODOs par our discussion

zheng-da committed:
handle the case that outputs are inputs.

zheng-da committed:
handle the case that inputs aren't used.

junrushao committed:
Add comments for _union_inputs

junrushao committed:
Make sure type/shape/storage type of loop_vars and new_loop_vars are consistent

junrushao committed:
Check consistency between inputs of main graph and inputs of subgraphs

junrushao committed:
Fix tests

junrushao committed:
Make testcases consistent with the new API

junrushao committed:
Fix typo

zheng-da committed:
handle the case without output data.

zheng-da committed:
fix a bug in foreach backward.

junrushao committed:
Group outputs into two lists

junrushao committed:
Enable testcase 2.3

junrushao committed:
Add _verify_while_loop

zheng-da committed:
fix a bug when there isn't output data.

junrushao committed:
...

junrushao committed:
Add first testcase

zheng-da committed:
Fix lint error.

zheng-da committed:
test diff Gluon RNN cells.

zheng-da committed:
test all symbol RNN cells.

junrushao committed:
Fix shape inference

junrushao committed:
Disable false assertion

junrushao committed:
Cleanup unittest

zheng-da committed:
adjust the test precision.

junrushao committed:
Simplify consistency check

junrushao committed:
Fix indexing issues in while's forward

junrushao committed:
Fix indexing in unittest

junrushao committed:
Fix indexing bugs in syncing inputs and outputs

zheng-da committed:
Fix a bug in getting a list of variable names.

We can't get a list of variable names from a hashtable. The order can't
be guaranteed. Python2 and Python3 output different orders.

zheng-da committed:
fix lint error.

junrushao committed:
Check return value of sync

junrushao committed:
A lots of testcases (in which I failed some in backwards...)

zheng-da committed:
Test 1D array.

zheng-da committed:
fix a bug when subgraph inputs and outputs share NDArray.

zheng-da committed:
handle the case that inputs and outputs of a graph share NDArrays

zheng-da committed:
add test.

zheng-da committed:
test multiple times.

zheng-da committed:
don't change the state's array list.

zheng-da committed:
trigger CI

zheng-da committed:
fix.

zheng-da committed:
fix

junrushao committed:
TO BE REVERTED!!!

junrushao committed:
Fix rebase error

junrushao committed:
tmp
